odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+                # The first occurs for `reduce` with all axes,
+                # the second for in-place stuff (`at` currently)
+                return res
+
+            # Wrap result if necessary (lazily)
+            if out is None:
+                if is_floating_dtype(res.dtype):
+                    if res.shape != self.shape:
+                        # Don't propagate weighting if shape changes
+                        weighting = CupyTensorSpaceConstWeighting(1.0,
+                                                                  exponent)
+                    spc_kwargs = {'weighting': weighting}
+                else:
+                    spc_kwargs = {}
+
+                res_space = type(self.space)(
+                    res.shape, res.dtype, self.device, **spc_kwargs)
+                result = res_space.element(res)
+            else:
+                result = out_tuple[0]
+
+            return result
+
+    @property
+    def ufuncs(self):
+        """Access to NumPy style ufuncs.
+
+        Examples
+        --------
+        >>> r2 = odl.rn(2, impl='cupy')
+        >>> x = r2.element([1, -2])
+        >>> x.ufuncs.absolute()
+        rn(2, impl='cupy').element([ 1.,  2.])
+
+        These functions can also be used with broadcasting or
+        array-like input:
+
+        >>> x.ufuncs.add(3)
+        rn(2, impl='cupy').element([ 4.,  1.])
+        >>> x.ufuncs.subtract([3, 3])
+        rn(2, impl='cupy').element([-2., -5.])
+
+        There is also support for various reductions
+        (sum, prod, amin, amax):
+
+        >>> x.ufuncs.sum()
+        -1.0
+        >>> x.ufuncs.prod()
+        -2.0
+
+        They also support an out parameter
+
+        >>> y = r2.element([3, 4])
+        >>> out = r2.element()
+        >>> result = x.ufuncs.add(y, out=out)
+        >>> result
+        rn(2, impl='cupy').element([ 4.,  2.])
+        >>> result is out
+        True
+
+        Notes
+        -----
+        Those ufuncs which are implemented natively on the GPU incur no
+        significant overhead. However, for missing functions, a fallback
+        Numpy implementation is used which causes significant overhead
+        due to data copies between host and device.
+        """
+        # TODO: Test with some native ufuncs, then remove this attribute
+        return super(CupyTensor, self).ufuncs
+
+    @property
+    def real(self):
+        """Real part of this tensor.
+
+        Returns
+        -------
+        real : `CupyTensor` view with real dtype
+            The real part of this tensor as an element of an `rn` space.
+        """
+        # Only real dtypes currently
+        return self
+
+    @real.setter
+    def real(self, newreal):
+        """Setter for the real part.
+
+        This method is invoked by ``tensor.real = other``.
+
+        Parameters
+        ----------
+        newreal : `array-like` or scalar
+            The new real part for this tensor.
+        """
+        self.real.data[:] = newreal
```

Comment:
I think `self.data.real[:]` would be slightly more efficient?

Reply:
Very possible.